// payload/dist/fields/performFieldOperations.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const errors_1 = require("../errors");
const sanitizeFallbackLocale_1 = __importDefault(require("../localization/sanitizeFallbackLocale"));
const traverseFields_1 = __importDefault(require("./traverseFields"));
const deepCopyObject_1 = __importDefault(require("../utilities/deepCopyObject"));
async function performFieldOperations(entityConfig, args) {
    const { data, originalDoc: fullOriginalDoc, docWithLocales, operation, hook, req, id, req: { payloadAPI, locale, }, overrideAccess, flattenLocales, unflattenLocales = false, showHiddenFields = false, isVersion = false, skipValidation = false, } = args;
    const fullData = (0, deepCopyObject_1.default)(data);
    const fallbackLocale = (0, sanitizeFallbackLocale_1.default)(req.fallbackLocale);
    let depth = 0;
    if (payloadAPI === 'REST' || payloadAPI === 'local') {
        depth = (args.depth || args.depth === 0) ? parseInt(String(args.depth), 10) : this.config.defaultDepth;
        if (depth > this.config.maxDepth)
            depth = this.config.maxDepth;
    }
    const currentDepth = args.currentDepth || 1;
    // Maintain a top-level list of promises
    // so that all async field access / validations / hooks
    // can run in parallel
    const validationPromises = [];
    const accessPromises = [];
    const relationshipPopulations = [];
    const hookPromises = [];
    const unflattenLocaleActions = [];
    const transformActions = [];
    const errors = [];
    // //////////////////////////////////////////
    // Entry point for field validation
    // //////////////////////////////////////////
    (0, traverseFields_1.default)({
        fields: entityConfig.fields,
        data: fullData,
        originalDoc: fullOriginalDoc,
        path: '',
        flattenLocales,
        locale,
        fallbackLocale,
        accessPromises,
        operation,
        overrideAccess,
        req,
        id,
        relationshipPopulations,
        depth,
        currentDepth,
        hook,
        hookPromises,
        fullOriginalDoc,
        fullData,
        validationPromises,
        errors,
        payload: this,
        showHiddenFields,
        unflattenLocales,
        unflattenLocaleActions,
        transformActions,
        docWithLocales,
        isVersion,
        skipValidation,
    });
    if (hook === 'afterRead') {
        transformActions.forEach((action) => action());
    }
    const hookResults = hookPromises.map((promise) => promise());
    await Promise.all(hookResults);
    const validationResults = validationPromises.map((promise) => promise());
    await Promise.all(validationResults);
    if (errors.length > 0) {
        throw new errors_1.ValidationError(errors);
    }
    if (hook === 'beforeChange') {
        transformActions.forEach((action) => action());
    }
    unflattenLocaleActions.forEach((action) => action());
    const accessResults = accessPromises.map((promise) => promise());
    await Promise.all(accessResults);
    const relationshipPopulationResults = relationshipPopulations.map((population) => population());
    await Promise.all(relationshipPopulationResults);
    return fullData;
}
exports.default = performFieldOperations;
